src/modules/credits/CreditsClient.ts

```diff
@@ -22,29 +22,105 @@ export class CreditsClient extends BaseClient {
     }
   }
 
+  /**
+   * Creates an SSE connection with reconnection logic
+   * @param address - The user's address
+   * @param onMessage - Callback function for received messages
+   * @param onError - Callback function for errors
+   * @returns EventSource with auto-reconnection capabilities
+   */
   createSSEConnection(
     address: string,
     onMessage: (data: CreditsResponse) => void,
     onError: (error: Event) => void
   ): EventSource {
-    const eventSource = new EventSource(
-      `${this.url}/users/${address}/credits/stream`
-    )
-
-    eventSource.onmessage = event => {
-      try {
-        const data = JSON.parse(event.data) as CreditsResponse
-        onMessage(data)
-      } catch (error) {
-        console.error('Error parsing SSE event data:', error)
+    let eventSource: EventSource | null = null
+    let retryCount = 0
+    const MAX_RETRIES = 5
+    const RETRY_DELAY_MS = 3000
+
+    // Create a wrapper object that appears as an EventSource to TypeScript
+    const wrapper = {} as EventSource
+
+    const connect = () => {
+      if (eventSource) {
+        // Close existing connection before creating a new one
+        eventSource.close()
+      }
+
+      eventSource = new EventSource(
+        `${this.url}/users/${address}/credits/stream`
+      )
+
+      // Copy all properties from the real EventSource to our wrapper
+      Object.defineProperties(
+        wrapper,
+        Object.getOwnPropertyDescriptors(eventSource)
+      )
+
+      eventSource.onopen = event => {
+        // Reset retry count on successful connection
+        retryCount = 0
+        console.log('SSE connection established')
+        // Forward the event if there's a listener on wrapper
+        if (wrapper.onopen) {
+          wrapper.onopen(event)
+        }
+      }
+
+      eventSource.onmessage = event => {
+        try {
+          const data = JSON.parse(event.data) as CreditsResponse
+          onMessage(data)
+          // Forward the event if there's a listener on wrapper
+          if (wrapper.onmessage) {
+            wrapper.onmessage(event)
+          }
+        } catch (error) {
+          console.error('Error parsing SSE event data:', error)
+        }
+      }
+
+      eventSource.onerror = error => {
+        console.error('SSE connection error:', error)
+
+        // Forward the event if there's a listener on wrapper
+        if (wrapper.onerror && eventSource) {
+          wrapper.onerror(error)
+        }
+
+        // Close the connection
+        if (eventSource) {
+          eventSource.close()
+        }
+
+        // Attempt to reconnect if under max retries
+        if (retryCount < MAX_RETRIES) {
+          retryCount++
+          console.log(
+            `Reconnecting SSE (attempt ${retryCount}/${MAX_RETRIES})...`
+          )
+          setTimeout(connect, RETRY_DELAY_MS)
```

Comment:
If we expose the close function, we might want to store the `setTimeout` identifier and clear it when calling close, what do you think about it?

Reply:
sure, done